publisher/internal/analytics/ethereum/types.go

```diff
@@ -18,42 +34,224 @@ type EventLog struct {
 	LowerTick    int
 	UpperTick    int
-	CurrentRatio float64
-	ValueAdded   float64
+	TxHash       string
 }
 
-type TokenAddition struct {
+type TokenTransaction struct {
 	repository.Token
 	Amount float64
 	Price  float64
 }
 
-func (intRec *LiquidityAddition) setLiquidityRecordFields() {
-	intRec.Token0Amount = intRec.Token0.Amount
-	intRec.Token1Amount = intRec.Token1.Amount
-	intRec.Token0PriceUsd = intRec.Token0.Price
-	intRec.Token1PriceUsd = intRec.Token1.Price
-	intRec.Token0Symbol = intRec.Token0.Symbol
-	intRec.Token1Symbol = intRec.Token1.Symbol
+func (rem Removal) SaveToDB(ts time.Time, a *Analytics) error {
+	removal := repository.Removal{
+		TimestampReceived: ts,
+		LPoolAddress:      rem.Address,
+		Token0Symbol:      rem.Token0.Symbol,
+		Token1Symbol:      rem.Token1.Symbol,
+		Token0Amount:      rem.Token0.Amount,
+		Token1Amount:      rem.Token1.Amount,
+		LowerRatio:        rem.LowerRatio,
+		UpperRatio:        rem.UpperRatio,
+		Token0PriceUsd:    rem.Token0.Price,
+		Token1PriceUsd:    rem.Token1.Price,
+		TxHash:            rem.TxHash,
+	}
+	return a.db.SaveRemoval(removal)
+}
+
+func (add Addition) SaveToDB(ts time.Time, a *Analytics) error {
+	addition := repository.Addition{
+		TimestampReceived: ts,
+		LPoolAddress:      add.Address,
+		Token0Symbol:      add.Token0.Symbol,
+		Token1Symbol:      add.Token1.Symbol,
+		Token0Amount:      add.Token0.Amount,
+		Token1Amount:      add.Token1.Amount,
+		LowerRatio:        add.LowerRatio,
+		UpperRatio:        add.UpperRatio,
+		Token0PriceUsd:    add.Token0.Price,
+		Token1PriceUsd:    add.Token1.Price,
+		TxHash:            add.TxHash,
+	}
+	return a.db.SaveAddition(addition)
+}
+
+func (p Position) CanPublish() bool {
+	if strings.EqualFold(p.Token0.Symbol, "") || strings.EqualFold(p.Token1.Symbol, "") {
+		log.Printf("SKIP - token symbol unknown. Tx: %s\n\n", p.TxHash)
+		return false
+	}
+	if p.LowerRatio == 0 && p.UpperRatio == 0 {
+		log.Printf("SKIP - actual ratio not calculated. Tx: %s\n\n", p.TxHash)
+		return false
+	}
+	if p.Token0.Amount == 0 && p.Token1.Amount == 0 {
+		log.Printf("SKIP - no tokens moved. Tx: %s\n\n", p.TxHash)
+		return false
+	}
+
+	return true
+}
+
+func (r Removal) ExtractFromEventLogs(eLog EventLog, a *Analytics) (Operation, error) {
+	liqPool := eLog.Address
+	addr0, addr1, found := a.db.GetPoolPairAddresses(liqPool)
+	if !found {
+		return Removal{}, fmt.Errorf("SKIP - liq. pool is unknown (removal). pool address: %s", liqPool)
+	}
+	token0, found0 := a.db.GetToken(addr0)
+	token1, found1 := a.db.GetToken(addr1)
+	if !found0 || !found1 {
+		return Removal{}, fmt.Errorf("SKIP - at least one token is unknown in liquidity removal. pool address: %s", liqPool)
+	}
+
+	_, token0HexAmount, token1HexAmount, err := splitLogDatatoHexStrings(eLog.Data)
+	if err != nil {
+		return Removal{}, err
+	}
+
+	remPosition := Position{
+		Token0: TokenTransaction{
+			Token:  token0,
+			Amount: a.convertTransferAmount(token0HexAmount, token0.Decimals),
```

Comment:
Since we also depend on some functionality provided by `Analytics` it seems logical to use it like it is used here and above, but a better way would be to declare an interface with `ConvertTransferAmount` and other needed methods, and have a field that points to an instance(`*Analytics` in this case) that implements them.

This will also require adding `NewAdditionOperation` and `NewRemovalOperation` functions in order to avoid forgetting setting required fields in respective fields.